src/sparseml/pytorch/utils/trainer.py

```diff
@@ -0,0 +1,562 @@
+    InceptionCrossEntropyLossWrapper,
+    default_device,
+    TensorBoardLogger,
+    PythonLogger,
+    model_to_device,
+    ModuleTrainer,
+    ModuleDeviceContext,
+    ModuleTester,
+    DEFAULT_LOSS_KEY,
+    get_prunable_layers,
+    tensor_sparsity,
+)
+from sparseml.utils import create_dirs
+from sparsezoo import Zoo
+import torch
+import logging
+
+from torch import Module
+from torch.utils.data import DataLoader, Dataset
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ImageClassificationTrainer:
+    "Utility class for training image classification models"
+
+    def __init__(
+            self,
+            checkpoint_path: str,
+            recipe_path: str,
+            dataset: str,
+            dataset_path: str,
+            train_batch_size: int,
+            test_batch_size: int,
+            eval_mode: bool = False,
+            local_rank: int = -1,
+            loader_num_workers: int = 4,
+            loader_pin_memory: bool = True,
+            pretrained: Union[str, bool] = True,
+            pretrained_dataset: Optional[str] = None,
+            init_lr: float = 1e-9,
+            optimizer_name: str = 'adam',
+            sparse_transfer_learn: bool = False,
+            use_mixed_precision: bool = False,
+            device: str = default_device(),
+            debug_steps: int = -1,
+            save_best_after: int = -1,
+            save_epochs: List[int] = [],
+            optim_kwargs: Optional[Dict[str, Any]] = None,
+            dataset_kwargs: Optional[Dict[str, Any]] = None,
+            model_kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        self._checkpoint_path = checkpoint_path
+        self._checkpoint = self._load_checkpoint()
+
+        self.arch_key = self._load_arch_key_from_checkpoint()
+        self.recipe_path = recipe_path
+
+        self.input_shape = ModelRegistry.input_shape(self.arch_key)
+        self.input_image_size = self.input_shape[1]  # assume shape [C, S, S]
+        self.train_batch_size = train_batch_size
+        self.test_batch_size = test_batch_size
+        self.loader_num_workers = loader_num_workers
+        self.loader_pin_memory = loader_pin_memory
+        self.pretrained = pretrained
+        self.pretrained_dataset = pretrained_dataset
+        self.dataset = dataset
+        self.dataset_path = dataset_path
+        self.local_rank = local_rank
+        self.eval_mode = eval_mode
+        self.optimizer_name = optimizer_name
+        self.init_lr = init_lr
+        self.sparse_transfer_learn = sparse_transfer_learn
+        self.use_mixed_precision = use_mixed_precision
+        self.device = device
+        self.debug_steps = debug_steps
+        self.save_best_after = save_best_after
+        self.save_epochs = save_epochs
+
+        self.dataset_kwargs = dataset_kwargs or {}
+        self.optim_kwargs = optim_kwargs or {}
+
+        self.world_size = int(os.environ.get("WORLD_SIZE", 1))
+        self.env_rank = int(os.environ.get("RANK", -1))
```

Comment:
DDP should be handled outside of the trainer, if possible, is there any reason it has to be in here? most of the rank specific calls should happen outside the trainer, right?